### ISSUE TITLE
An implementation without the need of iOS or macOS environment

### ISSUE THREAD

**Issue by AsuharietYgvar** [KhaosT/nhcalc#2]
First of all, thanks for sharing your findings of this private api!

I have created [AppleNeuralHash2ONNX](https://github.com/AsuharietYgvar/AppleNeuralHash2ONNX), which can convert Apple's model to ONNX, by doing reverse-engineering on this API.  It can be used to calculate neural hash with [onnxruntime](https://github.com/microsoft/onnxruntime), even on Linux.

One suggestion: The API returns 96-bit binary data encoded in base64. You can decode it to raw binary and encode again with hex. This will result in 24 hex characters instead of 32 produced here, making it easier to do bitwise comparisons.

**Comment by KhaosT** [comment]
That's really cool 🙏 Hopefully these tools will help researcher to better measure NeuralHash's collision rates.

**Comment by AsuharietYgvar** [comment]
Indeed. In theory you can train a GAN using the exported ONNX model as the discriminator. Then you can use the generator to generate a random image for any input hash. This can be used to spam Apple with endless false positives if we can also export the CSAM database on device.

**Comment by nicolas17** [comment]
> if we can also export the CSAM database on device

No you can't.

![image](https://user-images.githubusercontent.com/101955/129934863-862f4a4f-9281-4c41-8a3b-7983ab8e6c40.png)
